packages/xod-arduino/src/transpiler.js

```diff
@@ -341,42 +348,52 @@ const createTNodes = def(
  * and it has a ready-to-use values, nothing needed to compute anymore.
  */
 export const transformProject = def(
-  'transformProject :: [Source] -> Project -> PatchPath -> Either Error TProject',
-  (impls, project, path) => R.compose(
-    R.chain((tProject) => {
+  'transformProject :: [Source] -> Project -> PatchPath -> TranspilationOptions -> Either Error TProjectWithNodeIdsMap',
+  (impls, project, path, opts) => R.compose(
+    R.chain((data) => {
       const nodeWithTooManyOutputs = R.find(
         R.pipe(R.prop('outputs'), R.length, R.lt(7)),
-        tProject.patches
+        data.project.patches
       );
 
       if (nodeWithTooManyOutputs) {
         const { owner, libName, patchName } = nodeWithTooManyOutputs;
         return Either.Left(new Error(`Native node ${owner}/${libName}/${patchName} has more than 7 outputs`));
       }
 
-      return Either.of(tProject);
-    }),
-    R.map((proj) => {
-      const patches = createTPatches(path, proj);
-
-      return R.applySpec({
-        config: createTConfig(path),
-        patches: R.always(patches),
-        nodes: createTNodes(path, patches),
-      })(proj);
+      return Either.of(data);
     }),
+    R.map(R.evolve({
+      project: (proj) => {
+        const patches = createTPatches(path, proj);
+
+        return R.applySpec({
+          config: createTConfig(opts, path),
+          patches: R.always(patches),
+          nodes: createTNodes(path, patches),
+        })(proj);
+      },
+    })),
     R.chain(R.compose(
       toposortProject(path),
-      Project.extractBoundInputsToConstNodes(R.__, path, project)
+      Project.extractBoundInputsToConstNodes(R.__, path, project),
     )),
-    Project.flatten
-  )(project, path, impls)
+    R.chain(Project.flatten(R.__, path, impls)),
+    R.ifElse(
+      () => (opts.debug === true),
+      Either.of,
+      Project.updatePatch(path, Project.removeDebugNodes)
+    )
+  )(project)
 );
 
 export default def(
-  'transpile :: Project -> PatchPath -> Either Error String',
+  'transpile :: Project -> PatchPath -> TranspilationOptions -> Either Error TranspilationResult',
   R.compose(
-    R.map(renderProject),
+    R.map(R.applySpec({
+      code: R.pipe(R.prop('project'), renderProject),
+      nodeIdsMap: R.prop('nodeIdsMap'),
+    })),
```

Comment:
Instead of this ad-hoc object we could export the `transformProject` and `getOriginalIdMap :: TProject -> StrMap NodeId`